src/lightning/fabric/utilities/throughput_monitor.py

```diff
@@ -0,0 +1,367 @@
+            elapsed_time = self._time[-1] - self._time[0]
+            dev_samples_per_sec = elapsed_samples / elapsed_time
+            dev_batches_per_sec = elapsed_batches / elapsed_time
+            metrics.update(
+                {
+                    f"device{self.separator}batches_per_sec": elapsed_batches / elapsed_time,
+                    f"device{self.separator}samples_per_sec": dev_samples_per_sec,
+                }
+            )
+            if add_global_metrics:
+                samples_per_sec = dev_batches_per_sec * self.world_size
+                metrics.update(
+                    {
+                        "batches_per_sec": samples_per_sec,
+                        "samples_per_sec": dev_samples_per_sec * self.world_size,
+                    }
+                )
+            if lengths is not None:
+                elapsed_lengths = int(self._lengths[-1]) - int(self._lengths[0])
+                avg_length = elapsed_lengths / elapsed_batches
+                if add_global_metrics:
+                    metrics["items_per_sec"] = samples_per_sec * avg_length
+                metrics[f"device{self.separator}items_per_sec"] = dev_samples_per_sec * avg_length
+
+        if flops_per_batch is not None:
+            # sum of flops per batch across ranks
+            self._flops.append(flops_per_batch * self.world_size)
+        if len(self._flops) == self._flops.maxlen:
+            elapsed_flops = sum(self._flops) - self._flops[0]
+            elapsed_time = self._time[-1] - self._time[0]
+            flops_per_sec = elapsed_flops / elapsed_time
+            dev_flops_per_sec = flops_per_sec / self.world_size
+            if add_global_metrics:
+                metrics["flops_per_sec"] = flops_per_sec
+            metrics[f"device{self.separator}flops_per_sec"] = dev_flops_per_sec
+            if self.flops_available:
+                metrics[f"device{self.separator}mfu"] = dev_flops_per_sec / self.flops_available
+
+        return metrics
+
+
+class ThroughputMonitor:
+    r"""Tracks and logs throughput with the :class:`_ThroughputMonitor`
+
+    Args:
+        fabric: The Fabric object.
+        \**kwargs: See available parameters in :class:`_ThroughputMonitor`
+
+    """
+
+    def __init__(self, fabric: "Fabric", **kwargs: Any) -> None:
+        fabric._validate_launched()  # otherwise world_size might be incorrect
+        dtype = _plugin_to_compute_dtype(fabric.strategy.precision)
+        flops_available = _get_flops_available(fabric.device, dtype)
+        self._fabric = fabric
+        self.step = -1
+        self._monitor = _ThroughputMonitor(flops_available=flops_available, world_size=fabric.world_size, **kwargs)
+
+    @rank_zero_only
+    def compute(self, *args: Any, **kwargs: Any) -> None:
+        """See :meth:`_ThroughputMonitor.compute`"""
+        self.step += 1
+        metrics = self._monitor.compute(*args, **kwargs)
+        self._fabric.log_dict(metrics=metrics, step=self.step)
```

Comment:
A method/function that is called compute but doesn't return anything is odd in my opinion. Also, in reality it actually computes and logs at the same time. I still think this should return the metrics, because otherwise I have no control over how things get logged, and the step is internally set. This is undesirable IMO. What I'd like to do is collect all metrics under one step, instead of logging to multiple steps:

```py
# put everything in a dict
metrics = {"loss": loss}
metrics = {"accuracy: acc}
metrics.update(monitor.compute())

# now log everything together
fabric.log_dict(metrics, step)
```

Reply:
Supported now